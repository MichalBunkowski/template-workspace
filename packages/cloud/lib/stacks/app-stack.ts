import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';

import { CognitoConstruct } from '../features/auth/cognito.construct';
import { AmplifyConstruct } from '../features/deployment/amplify.construct';
import { CommonProps } from '../types/interfaces/common-props';
import generateResourceName from '../utils/generate-resource-name.utils';

type AppStackProps = CommonProps;

export class AppStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: AppStackProps) {
    super(scope, id, props);

    // Authentication
    /**
     *  CognitoConstruct
     *  Purpose: Authentication of users and confirming identity
     * */
    const cognito = new CognitoConstruct(
      this,
      generateResourceName('Auth', props),
      {
        ...props,
      }
    );

    // Deployment
    /**
     *  amplifyServiceRole
     *  Purpose: Allows amplify app to create and deploy backend resources
     * */
    const amplifyServiceRole = new iam.Role(
      this,
      generateResourceName('AmplifyBackendDeployment', props),
      {
        assumedBy: new iam.ServicePrincipal('amplify.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'AdministratorAccess-Amplify'
          ),
        ],
      }
    );

    /**
     *  githubSecret
     *  Purpose: Access to Github repository
     * */
    const githubToken = cdk.SecretValue.secretsManager('github_token');

    /**
     *  AmplifyConstruct
     *  Purpose: Deployment of application
     * */
    new AmplifyConstruct(this, generateResourceName('Amplify', props), {
      serviceRole: amplifyServiceRole,
      githubToken: githubToken,
      cognito,
      ...props,
    });
  }
}
