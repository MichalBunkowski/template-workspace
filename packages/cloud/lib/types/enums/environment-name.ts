export enum EnvironmentName {
  Develop = 'deve',
  Production = 'prod',
}
