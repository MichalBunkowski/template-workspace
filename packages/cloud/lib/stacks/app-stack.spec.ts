import * as cdk from '@aws-cdk/core';

import { EnvironmentName } from '../types/enums/environment-name';
import { AppStack } from './app-stack';

describe('AppStack', () => {
  it('should match snapshot', () => {
    const app = new cdk.App();
    // WHEN
    const stack = new AppStack(app, 'MyTestStack', {
      envName: EnvironmentName.Develop,
    });
    // THEN
    const actual = app.synth().getStackArtifact(stack.artifactId).template;
    expect(actual.Resources ?? {}).toMatchSnapshot();
  });
});
