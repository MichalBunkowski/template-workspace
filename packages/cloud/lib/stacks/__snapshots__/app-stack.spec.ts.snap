// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppStack should match snapshot 1`] = `
Object {
  "AmplifyAmplifyApp0346D4EC": Object {
    "Properties": Object {
      "BasicAuthConfig": Object {
        "EnableBasicAuth": false,
      },
      "EnableBranchAutoDeletion": true,
      "EnvironmentVariables": Array [
        Object {
          "Name": "_LIVE_UPDATES",
          "Value": "[{\\"pkg\\":\\"next-version\\",\\"type\\":\\"internal\\",\\"version\\":\\"latest\\"}]",
        },
        Object {
          "Name": "AMPLIFY_MONOREPO_APP_ROOT",
          "Value": "packages/web",
        },
        Object {
          "Name": "AMPLIFY_DIFF_DEPLOY",
          "Value": "false",
        },
        Object {
          "Name": "NEXT_CLIENT_REGION",
          "Value": "eu-west-1",
        },
        Object {
          "Name": "NEXT_CLIENT_IDENTITY_POOL_ID",
          "Value": Object {
            "Ref": "AuthAuthIdentityPool71ACCEC7",
          },
        },
        Object {
          "Name": "NEXT_CLIENT_USER_POOL_ID",
          "Value": Object {
            "Ref": "AuthAuthUserPool2E5A325E",
          },
        },
        Object {
          "Name": "NEXT_CLIENT_USER_POOL_CLIENT_ID",
          "Value": Object {
            "Ref": "AuthAuthUserPoolClient493875FB",
          },
        },
      ],
      "IAMServiceRole": Object {
        "Fn::GetAtt": Array [
          "AmplifyBackendDeployment84142950",
          "Arn",
        ],
      },
      "Name": "AmplifyApp",
      "OauthToken": "{{resolve:secretsmanager:github_token:SecretString:::}}",
      "Repository": "https://github.com/MichalBunkowski/template-workspace",
    },
    "Type": "AWS::Amplify::App",
  },
  "AmplifyAmplifyAppmain87A1DA5F": Object {
    "Properties": Object {
      "AppId": Object {
        "Fn::GetAtt": Array [
          "AmplifyAmplifyApp0346D4EC",
          "AppId",
        ],
      },
      "BranchName": "main",
      "EnableAutoBuild": false,
      "EnablePullRequestPreview": true,
    },
    "Type": "AWS::Amplify::Branch",
  },
  "AmplifyBackendDeployment84142950": Object {
    "Properties": Object {
      "AssumeRolePolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": "sts:AssumeRole",
            "Effect": "Allow",
            "Principal": Object {
              "Service": "amplify.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
      "ManagedPolicyArns": Array [
        Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/AdministratorAccess-Amplify",
            ],
          ],
        },
      ],
    },
    "Type": "AWS::IAM::Role",
  },
  "AuthAuthIdentityPool71ACCEC7": Object {
    "Properties": Object {
      "AllowUnauthenticatedIdentities": false,
      "CognitoIdentityProviders": Array [
        Object {
          "ClientId": Object {
            "Ref": "AuthAuthUserPoolClient493875FB",
          },
          "ProviderName": Object {
            "Fn::GetAtt": Array [
              "AuthAuthUserPool2E5A325E",
              "ProviderName",
            ],
          },
        },
      ],
    },
    "Type": "AWS::Cognito::IdentityPool",
  },
  "AuthAuthRoleConstructAuthIdentityPoolAttachment7C8D5F89": Object {
    "Properties": Object {
      "IdentityPoolId": Object {
        "Ref": "AuthAuthIdentityPool71ACCEC7",
      },
      "Roles": Object {
        "authenticated": Object {
          "Fn::GetAtt": Array [
            "AuthAuthRoleConstructAuthRoleFB3DB2BB",
            "Arn",
          ],
        },
      },
    },
    "Type": "AWS::Cognito::IdentityPoolRoleAttachment",
  },
  "AuthAuthRoleConstructAuthRoleDefaultPolicy3F517683": Object {
    "Properties": Object {
      "PolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": Array [
              "mobileanalytics:PutEvents",
              "cognito-sync:*",
              "cognito-identity:*",
            ],
            "Effect": "Allow",
            "Resource": "*",
          },
        ],
        "Version": "2012-10-17",
      },
      "PolicyName": "AuthAuthRoleConstructAuthRoleDefaultPolicy3F517683",
      "Roles": Array [
        Object {
          "Ref": "AuthAuthRoleConstructAuthRoleFB3DB2BB",
        },
      ],
    },
    "Type": "AWS::IAM::Policy",
  },
  "AuthAuthRoleConstructAuthRoleFB3DB2BB": Object {
    "Properties": Object {
      "AssumeRolePolicyDocument": Object {
        "Statement": Array [
          Object {
            "Action": "sts:AssumeRoleWithWebIdentity",
            "Condition": Object {
              "ForAnyValue:StringLike": Object {
                "cognito-identity.amazonaws.com:amr": "authenticated",
              },
              "StringEquals": Object {
                "cognito-identity.amazonaws.com:aud": Object {
                  "Ref": "AuthAuthIdentityPool71ACCEC7",
                },
              },
            },
            "Effect": "Allow",
            "Principal": Object {
              "Federated": "cognito-identity.amazonaws.com",
            },
          },
        ],
        "Version": "2012-10-17",
      },
    },
    "Type": "AWS::IAM::Role",
  },
  "AuthAuthUserPool2E5A325E": Object {
    "DeletionPolicy": "Retain",
    "Properties": Object {
      "AccountRecoverySetting": Object {
        "RecoveryMechanisms": Array [
          Object {
            "Name": "verified_email",
            "Priority": 1,
          },
        ],
      },
      "AdminCreateUserConfig": Object {
        "AllowAdminCreateUserOnly": false,
      },
      "AliasAttributes": Array [
        "email",
      ],
      "AutoVerifiedAttributes": Array [
        "email",
      ],
      "EmailVerificationMessage": "The verification code to your new account is {####}",
      "EmailVerificationSubject": "Verify your new account",
      "Policies": Object {
        "PasswordPolicy": Object {
          "MinimumLength": 8,
          "RequireLowercase": true,
          "RequireNumbers": true,
          "RequireSymbols": true,
          "RequireUppercase": true,
          "TemporaryPasswordValidityDays": 1,
        },
      },
      "Schema": Array [
        Object {
          "Mutable": false,
          "Name": "email",
          "Required": true,
        },
      ],
      "SmsVerificationMessage": "The verification code to your new account is {####}",
      "VerificationMessageTemplate": Object {
        "DefaultEmailOption": "CONFIRM_WITH_CODE",
        "EmailMessage": "The verification code to your new account is {####}",
        "EmailSubject": "Verify your new account",
        "SmsMessage": "The verification code to your new account is {####}",
      },
    },
    "Type": "AWS::Cognito::UserPool",
    "UpdateReplacePolicy": "Retain",
  },
  "AuthAuthUserPoolClient493875FB": Object {
    "Properties": Object {
      "AllowedOAuthFlows": Array [
        "implicit",
        "code",
      ],
      "AllowedOAuthFlowsUserPoolClient": true,
      "AllowedOAuthScopes": Array [
        "profile",
        "phone",
        "email",
        "openid",
        "aws.cognito.signin.user.admin",
      ],
      "CallbackURLs": Array [
        "https://example.com",
      ],
      "GenerateSecret": false,
      "SupportedIdentityProviders": Array [
        "COGNITO",
      ],
      "UserPoolId": Object {
        "Ref": "AuthAuthUserPool2E5A325E",
      },
    },
    "Type": "AWS::Cognito::UserPoolClient",
  },
  "AuthUserPoolDomain97947401": Object {
    "Properties": Object {
      "Domain": "home-assistance",
      "UserPoolId": Object {
        "Ref": "AuthAuthUserPool2E5A325E",
      },
    },
    "Type": "AWS::Cognito::UserPoolDomain",
  },
}
`;
